### PR TITLE
Explicit content part templating

### LINKED ISSUE #2363: Alternates for Dynamic Parts Not Added to Shape Metadata
Content [part templates](https://orchardcore.readthedocs.io/en/latest/OrchardCore.Modules/OrchardCore.Templates/README/#content-part-templates) are not available to dynamically defined parts (a dynamically defined part doesn't make use of the `ContentPartDisplayDriver<TPart>`, which is currently responsible for adding the necessary alternates). 

It would be useful (and less misleading for new comers) if [these templates](https://orchardcore.readthedocs.io/en/latest/OrchardCore.Modules/OrchardCore.Templates/README/#content-part-templates) were available by default for dynamically defined parts as well as strongly typed parts (those with corresponding C# classes).

Do others agree that this is first a problem? and that it is worth solving?

### Possible Approach

The solution seems to require a default ContentPartDisplayDriver which operates on the basic `ContentPart` object (as opposed to ContentPart subclasses). `ContentItemDisplayCoordinator.BuildDisplayAsync` is responsible for looping through the part drivers that have been registered, and calling every driver's `BuildDisplayAsync` for every part. 
```
    foreach (var displayDriver in _partDisplayDrivers)
    {
        try
        {
            var result = await displayDriver.BuildDisplayAsync(part, contentTypePartDefinition, context);
            if (result != null)
            {
                await result.ApplyAsync(context);
            }
       ...
```

The part drivers' `BuildDisplayAsync` methods are responsible for checking if they should return early (because the part argument passed in is not of the correct type). i.e. the first few lines look like this 
```
    var part = contentPart as TPart;
    if (part == null)
    {
        return Task.FromResult<IDisplayResult>(null);
    }
```

#### Implementing a Solution

If a default ContentPartDisplayDriver would be implemented, inheriting the existing `ContentPartDisplayDriver<TPart>` (where TPart : ContentPart), it could override the BuildDisplayAsync method and perform a more strict type check before returning early (else calling the base implementation) i.e.
```
    if (part.GetType() != typeof(ContentPart))
    {
        return Task.FromResult<IDisplayResult>(null);
    }
```
This way this default driver would only operate on parts that are not subclasses of ContentPart (indicating that they are dynamically defined in admin UI because they do not have their own ContentPart subclass defined in C#).

I'm not sure if other changes would be required to support this, and I haven't tried it myself yet. It's just a theory before I shoot off for work.

cc @sebastienros, @jtkech (as per our brief discussion on gitter)

### PR DESCRIPTION
**Pull request by aodl**
fixes #2363 

See #2363 for a description of the problem.

Fix description:
- Provide a default content part display driver which handles content parts that have no specialised C# model, just the standard ContentPart class, meaning that the content part was dynamically defined via the admin dashboard.
  - Ignore **implicit content parts** (because the **fields** conceptually belong to the content item, not the part, so templating should be provided at the content item / zone level)

- Make sure the **fields** for **explicit content parts** get added to the part shape, rather than the zone / holding shape.
  - This is achieved by getting BuildFieldDisplayContext to implement a simple IFieldDisplayManagementContext interface, which the ShapeResult can use to determine whether or not the shape represents a field of an explicit content part or not (so it knows which parent shape to add the field shape to - content part or zone holding)

- Finally provide a default ContentPart template (in case the user doesn't define one for their content part from the admin dashboard) and an ExplicitContentPart model that makes the meaning clearer (just in terms of iterating the fields property and calling display on each field shape)

### CODE REVIEW

File: src/OrchardCore/OrchardCore.ContentManagement.Display/ContentDisplay/DefaultContentPartDisplayDriver.cs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+using OrchardCore.ContentManagement.Metadata;
+using OrchardCore.ContentManagement.Metadata.Models;
+using OrchardCore.DisplayManagement.Handlers;
+using OrchardCore.DisplayManagement.Views;
+
+namespace OrchardCore.ContentManagement.Display.ContentDisplay
+{
+    /// <summary>
+    /// 
+    /// </summary>
```
Comment (Jetski5822) [inline src/OrchardCore/OrchardCore.ContentManagement.Display/ContentDisplay/DefaultContentPartDisplayDriver.cs]: ditch empty summary tags

File: src/OrchardCore/OrchardCore.ContentManagement.Display/ViewModels/ContentPartViewModel.cs [hunk 2]
```diff
@@ -15,4 +18,21 @@ public ContentPartViewModel(ContentPart contentPart)
 
         public ContentPart ContentPart { get; set; }
     }
+
+    // Models an explicit part (part name != content type name)
+    public class ExplicitContentPartViewModel : Shape
+    {
+        public ExplicitContentPartViewModel()
+        {
+        }
```
Comment (Jetski5822) [inline src/OrchardCore/OrchardCore.ContentManagement.Display/ViewModels/ContentPartViewModel.cs]: remove

Comment (aodl) [reply src/OrchardCore/OrchardCore.ContentManagement.Display/ViewModels/ContentPartViewModel.cs]: I was copying the fact that ContentPartViewModel had an explicit parameterless constructor for consistency. I've removed both of those now

**Comment by aodl** [comment]
@sebastienros , have you managed to find the time to take a look at this PR yet? It's only that with this branch hanging around it's starting to encounter merge conflicts with the dev branch (just a minor one at the moment).

If its the case that you're thinking about a better way of solving this issue, are you able to discuss concern's that you have with this approach?

**Comment by sebastienros** [comment]
Currently working on this PR, are you trying to solve #2363 or does it do more than just that?

**Pull request closed without merging**